openmeter/ent/schema/billing.go

```diff
@@ -0,0 +1,418 @@
+			})
+		case provider.TypeStripe:
+			return json.Marshal(providerConfigSerde[provider.StripeConfig]{
+				Meta:   provider.Meta{Type: provider.TypeStripe},
+				Config: config.Stripe,
+			})
+		default:
+			return nil, fmt.Errorf("unknown backend type: %s", config.Type)
+		}
+	},
+	S: func(ns *sql.NullString) (provider.Configuration, error) {
+		if !ns.Valid {
+			return provider.Configuration{}, errors.New("backend config is null")
+		}
+
+		data := []byte(ns.String)
+
+		var meta provider.Meta
+		if err := json.Unmarshal(data, &meta); err != nil {
+			return provider.Configuration{}, err
+		}
+
+		switch meta.Type {
+		case provider.TypeOpenMeter:
+			serde := providerConfigSerde[provider.OpenMeterConfig]{}
+
+			if err := json.Unmarshal(data, &serde); err != nil {
+				return provider.Configuration{}, err
+			}
+
+			return provider.Configuration{
+				Meta:      serde.Meta,
+				OpenMeter: serde.Config,
+			}, nil
+		case provider.TypeStripe:
+			serde := providerConfigSerde[provider.StripeConfig]{}
+
+			if err := json.Unmarshal(data, &serde); err != nil {
+				return provider.Configuration{}, err
+			}
+
+			return provider.Configuration{
+				Meta:   serde.Meta,
+				Stripe: serde.Config,
+			}, nil
+		default:
+			return provider.Configuration{}, fmt.Errorf("unknown backend type: %s", meta.Type)
+		}
+	},
+}
+
+type BillingWorkflowConfig struct {
+	ent.Schema
+}
+
+func (BillingWorkflowConfig) Mixin() []ent.Mixin {
+	return []ent.Mixin{
+		entutils.IDMixin{},
+		entutils.NamespaceMixin{},
+		entutils.TimeMixin{},
+	}
+}
+
+func (BillingWorkflowConfig) Fields() []ent.Field {
+	return []ent.Field{
+		field.Enum("alignment").
+			GoType(billing.AlignmentKind("")),
+
+		// TODO: later we will add more alignment details here (e.g. monthly, yearly, etc.)
+
+		field.Int64("collection_period_seconds"),
+
+		field.Bool("invoice_auto_advance").
+			Nillable(),
+
+		field.Int64("invoice_draft_period_seconds"),
+
+		field.Int64("invoice_due_after_seconds"),
+
+		field.Enum("invoice_collection_method").
+			GoType(billing.CollectionMethod("")),
+
+		field.Enum("invoice_line_item_resolution").
+			GoType(billing.GranualityResolution("")),
+
+		field.Bool("invoice_line_item_per_subject").
+			Default(false),
+	}
+}
+
+func (BillingWorkflowConfig) Indexes() []ent.Index {
+	return []ent.Index{
+		index.Fields("namespace", "id"),
+	}
+}
+
+func (BillingWorkflowConfig) Edges() []ent.Edge {
+	return []ent.Edge{
+		edge.To("billing_invoices", BillingInvoice.Type),
+		edge.To("billing_profile", BillingProfile.Type),
+	}
+}
+
+type BillingInvoiceItem struct {
+	ent.Schema
+}
+
+func (BillingInvoiceItem) Mixin() []ent.Mixin {
+	return []ent.Mixin{
+		entutils.IDMixin{},
+		entutils.NamespaceMixin{},
+		entutils.TimeMixin{},
+		entutils.MetadataAnnotationsMixin{},
+	}
+}
+
+func (BillingInvoiceItem) Fields() []ent.Field {
+	return []ent.Field{
+		field.String("invoice_id").
+			Optional().
+			SchemaType(map[string]string{
+				"postgres": "char(26)",
```

Comment:
nit: same, mixin?

Reply:
The invoice ID is used only once. I wouldn't want to create a parameterized mixin for ulid.